File: JS/webgl_objects/webglobject.js
import { addObject, removeObject } from "../webgl_data/glscene.js";
import { setScale, setAngleDegreesZ, setAngleDegrees, degreesToRadians, setAngleZ, setAngle, setPositionXY, setPosition, delete$ } from "../webgl_data/globj.js";
import { class_type } from "../.fable/fable-library.3.0.0/Reflection.js";
import { Vec2_Create_7B00E9A0 } from "../core/vectors.js";
import { DEG_PER_RAD } from "../core/utils.js";

export class WebglObject {
    constructor(scene, objectCreator) {
        this.scene = scene;
        this.objectDef = addObject(objectCreator)(this.scene);
    }
    ["Wil.Webgl.WebglObject.Delete"]() {
        const _ = this;
        const value = removeObject(_.objectDef)(_.scene);
        void value;
        delete$(_.objectDef);
    }
}

export function WebglObject$reflection() {
    return class_type("Wil.Webgl.WebglObject", void 0, WebglObject);
}

export function WebglObject_$ctor_E676F7D(scene, objectCreator) {
    return new WebglObject(scene, objectCreator);
}

export function WebglObject__get_ObjectDef(_) {
    return _.objectDef;
}

export function WebglObject__get_Scene(_) {
    return _.objectDef.Scene;
}

export function WebglObject__get_Position(_) {
    return _.objectDef.Position;
}

export function WebglObject__set_Position_Z3D47FC51(_, value) {
    setPosition(value, _.objectDef);
}

export function WebglObject__get_Position2(_) {
    const __1 = _.objectDef.Position;
    return Vec2_Create_7B00E9A0(__1.v.contents.values[0], __1.v.contents.values[1]);
}

export function WebglObject__set_Position2_Z3D47FC52(_, value) {
    setPositionXY(value.values[0], value.values[1], _.objectDef);
}

export function WebglObject__get_Angle(_) {
    return _.objectDef.Angle;
}

export function WebglObject__set_Angle_Z3D47FC51(_, value) {
    setAngle(value, _.objectDef);
}

export function WebglObject__get_AngleZ(_) {
    return _.objectDef.Angle.values[2] * 1;
}

export function WebglObject__set_AngleZ_6069AC9A(_, value) {
    setAngleZ(value, _.objectDef);
}

export function WebglObject__get_AngleDegrees(_) {
    return degreesToRadians(_.objectDef.Angle);
}

export function WebglObject__set_AngleDegrees_Z3D47FC51(_, value) {
    setAngleDegrees(value, _.objectDef);
}

export function WebglObject__get_AngleDegreesZ(_) {
    return (_.objectDef.Angle.values[2] * 1) * DEG_PER_RAD;
}

export function WebglObject__set_AngleDegreesZ_60759B6B(_, value) {
    setAngleDegreesZ(value, _.objectDef);
}

export function WebglObject__get_Scale(_) {
    return _.objectDef.Scale;
}

export function WebglObject__set_Scale_5E38073B(_, value) {
    setScale(value, _.objectDef);
}

File: src/webgl_objects/webglobject.fs
namespace Wil.Webgl

open Wil.Core
open Wil.Core.Utils
open Wil.Webgl.Data

module WebglObject =
  let inline get (name: string) uniforms =
    match uniforms |> Map.tryFind name with
    | Some uniform -> uniform
    | None -> throw $"Uniform '{name}' not found."

  let inline getInt uniforms name = ((get name uniforms).Value |> asArray<int>).[0]
  let inline setInt uniforms name (value: int) = get name uniforms |> GlUniform.setValue value
  let inline getFloat uniforms name = ((get name uniforms).Value |> asArray<float>).[0]
  let inline setFloat uniforms name (value: float) = get name uniforms |> GlUniform.setValue value
  let inline getVec2 uniforms name = Vec2((get name uniforms).Value |> asArray<float>)
  let inline setVec2 uniforms name (value: Vec2) = get name uniforms |> GlUniform.setValue value.Values
  let inline getVec3 uniforms name = Vec3((get name uniforms).Value |> asArray<float>)
  let inline setVec3 uniforms name (value: Vec3) = get name uniforms |> GlUniform.setValue value.Values
  let inline getVec4 uniforms name = Vec4((get name uniforms).Value |> asArray<float>)
  let inline setVec4 uniforms name (value: Vec4) = get name uniforms |> GlUniform.setValue value.Values
  let inline getIntVec2 uniforms name = let value = (get name uniforms).Value |> asArray<int> in Vec2.Create(float value.[0], float value.[1])
  let inline setIntVec2 uniforms name (value: Vec2) = get name uniforms |> GlUniform.setValue value.IntValues
  let inline getIntVec3 uniforms name = Vec3((get name uniforms).Value |> asArray<float>)
  let inline setIntVec3 uniforms name (value: Vec3) = get name uniforms |> GlUniform.setValue value.IntValues
  let inline getIntVec4 uniforms name = Vec4((get name uniforms).Value |> asArray<float>)
  let inline setIntVec4 uniforms name (value: Vec4) = get name uniforms |> GlUniform.setValue value.IntValues

  let inline pixelsToWorld (pixels: float<px>) scene =
    scene |> GlScene.pixelsToWorld pixels |> box

  let inline worldToPixels value scene =
    scene |> GlScene.worldToPixels value |> box

type WebglObject(scene: GlSceneData, objectCreator: GlObjectFactory) =
  let objectDef = GlScene.addObject objectCreator scene

  member _.ObjectDef with get() = objectDef
  member _.Scene with get() = objectDef.Scene

  member _.Position
    with get() = objectDef.Position
    and set(value) = objectDef |> GlObj.setPosition value

  member _.Position2
    with get() = objectDef.Position.XY
    and set(value: Vec2) = objectDef |> GlObj.setPositionXY value.X value.Y

  member _.Angle
    with get() = objectDef.Angle
    and set(value) = objectDef |> GlObj.setAngle value

  member _.AngleZ
    with get() = objectDef.Angle.Z |> radians
    and set(value) = objectDef |> GlObj.setAngleZ value

  member _.AngleDegrees
    with get() = GlObj.degreesToRadians objectDef.Angle
    and set(value) = objectDef |> GlObj.setAngleDegrees value

  member _.AngleDegreesZ
    with get() = objectDef.Angle.Z |> radians |> toDegrees
    and set(value) = objectDef |> GlObj.setAngleDegreesZ value

  member _.Scale
    with get() = objectDef.Scale
    and set(value) = objectDef |> GlObj.setScale value

  abstract member Delete: unit -> unit
  default _.Delete() =
    scene |> GlScene.removeObject objectDef |> ignore
    GlObj.delete objectDef
